File: cpp/benchmarks/gmodload.cpp
/**
 * @file gmodload.cpp
 * @brief Nanobench performance benchmark for Gmod singleton cache access
 * @details Measures the cost of retrieving an already-loaded Gmod from the VIS singleton cache.
 *          The Gmod JSON is loaded once at first access. Subsequent calls return the cached instance.
 *          This benchmark targets the cache-hit path, not JSON parsing.
 */

#define ANKERL_NANOBENCH_IMPLEMENT
#include <nanobench.h>

#include <dnv/VistaSDK.h>

namespace dnv::vista::sdk::benchmarks
{
    int run()
    {
        {
            const auto& warmup = VIS::instance().gmod(VisVersion::v3_7a);
            ankerl::nanobench::doNotOptimizeAway(&warmup);
        }

        ankerl::nanobench::Bench bench;
        bench.title("GmodCacheAccess").warmup(10000).minEpochIterations(1000000);

        bench.run("GmodCacheAccess_v3_4a", [&] {
            const auto& gmod = VIS::instance().gmod(VisVersion::v3_4a);
            ankerl::nanobench::doNotOptimizeAway(&gmod);
        });

        bench.run("GmodCacheAccess_v3_7a", [&] {
            const auto& gmod = VIS::instance().gmod(VisVersion::v3_7a);
            ankerl::nanobench::doNotOptimizeAway(&gmod);
        });

        return 0;
    }
} // namespace dnv::vista::sdk::benchmarks

int main()
{
    return dnv::vista::sdk::benchmarks::run();
}

File: bindings/rust/vista-sdk/benches/gmod_load.rs
use criterion::{criterion_group, criterion_main, Criterion};
use vista_sdk::core::vis::Vis;
use vista_sdk::core::vis_version::VisVersion;

fn bench(c: &mut Criterion) {
    let vis = Vis::instance();

    let _ = vis.gmod(VisVersion::V3_4a).expect("V3_4a should be valid");
    let _ = vis.gmod(VisVersion::V3_7a).expect("V3_7a should be valid");

    c.bench_function("GmodCacheAccess_v3_4a", |b| {
        b.iter(|| vis.gmod(VisVersion::V3_4a).expect("V3_4a should be valid"))
    });

    c.bench_function("GmodCacheAccess_v3_7a", |b| {
        b.iter(|| vis.gmod(VisVersion::V3_7a).expect("V3_7a should be valid"))
    });
}

criterion_group!(benches, bench);
criterion_main!(benches);
